scripts/scr_company_order/scr_company_order.gml

```diff
@@ -105,15 +108,103 @@ function scr_company_order(company) {
 		"Flash Git",
 		"Ork Sniper"
 	]
+
 	var role_shuffle_length = array_length(role_orders);
 	var company_length = array_length(name[co]);
+	var squadless={};
+	// find units not in a squad
+	for (i=0;i<company_length;i++;){
+		unit = TTRPG[co,i];
+		if (unit.squad=="none") and (unit.name()!=""){
+			if (!struct_exists(squadless, unit.role)){
+				squadless[$ unit.role] =[i];
+			} else {
+				array_push(squadless[$ unit.role],i)
+			}
+		}
+	}
+
+	//at this point check that all squads have the right types and numbers of units in them
+	var squad, wanted_roles;
+	for (i=0;i<array_length(squads);i++;){
+		if (squads[i].base_company != company) then continue;
+		squad = squads[i];
+		squad.update_fulfilment();
+		if (squad.fulfilled == false){
+			wanted_roles=struct_get_names(squad.required);
+
+			/* this finds sqauds that are in need of members and checks ot see if there 
+				are any squadless units in the chapter with
+				the rigth role to fill the gap*/ 
+			for (var r = 0;r < array_length(wanted_roles);r++;){
+				if (struct_exists(squadless,wanted_roles[r])){
+					while(array_length(squadless[$ wanted_roles[r]] > 0)) and (squad.required[$ wanted_roles[r]] > 0){
+						array_push(squad.members,[company,squadless[$ wanted_roles[r]][0]]);
+						array_delete(squadless[$ wanted_roles[r]],0,1);
+						squad.required[$ wanted_roles[r]]--;
+					}
+				}
+			}
+			//if no new sergeants are found for squad someon gets promoted
+			//find a new_sergeant 
+			if (struct_exists(squad.required, role[100,18])){
+				if (squad.required[$ role[100,18]] > 0){
+					squad.new_sergeant();
+					squad.required[$ role[100,18]]--;
+				}
+			}
+			//find a new veteran sergeant 
+			if (struct_exists(squad.required, role[100,19])){
+				if (squad.required[$ role[100,19]] > 0){
+					squad.new_sergeant(true);
+					squad.required[$ role[100,19]]--;
+				}
+			}		
+		}
+	}
+	var sorted_numbers = [];
+
+	//this stops over strenuous repeats should greatly speed up company reshuffle
+	for (i=0;i<company_length;i++;){
+		sorted_numbers[i]=i;
+	}
 	for (var role_name=0;role_name<role_shuffle_length;role_name++;){
```

Comment:
is this triple loop really necessary?

Reply:
> is this triple loop really necessary?

well before i did the refactor this script used to do 27 consecutive loops so comparably this is moving at light speed

Reply:
sweet lord

Reply:
i'll take a look through but i'm pretty sure this is as reduced down as i cant get it, but given it used to do a full loop over every role and then compare that role against every single unit slot even if it didn't hold unit data or had already matched , i feel i've probably freed up the processing to indulge in the loops.

Reply:
ok, let's leave it for now, if performance becomes an issue, we can go back to this